ee/clickhouse/timer.py

```diff
@@ -0,0 +1,105 @@
+import logging
+import uuid
+from collections import OrderedDict
+from functools import partial
+from threading import Condition, Thread
+from time import time
+from typing import Callable, Dict, Optional, Tuple
+
+from django.conf import settings
+
+logger = logging.getLogger(__name__)
+
+
+class TimerTask:
+    id: str
+    done: bool
+
+    def __init__(self, callback: Callable, *args, **kwargs):
+        self.callback = partial(callback, *args, **kwargs)
+        self.id = str(uuid.uuid4())
+        self.done = False
+
+    def run(self):
+        self.done = True
+        try:
+            self.callback()
+        except Exception as err:
+            logger.warn("TimerTask failed, ignoring error", err)
+
+
+class SingleThreadedTimer(Thread):
+    def __init__(self, timeout_ms: int, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.setDaemon(True)
+
+        self.timeout_ms = timeout_ms
+        self.started = False
+        self.lock = Condition()
+        self.tasks: OrderedDict = OrderedDict()
+
+    def schedule(self, callback: Callable, *args, **kwargs) -> TimerTask:
+        self.start()
```

Comment:
To answer your question about comments, a comment here would be nice:

"""
Schedules a `TimerTask` with a thread lock, noting the current time. See `run()` for consuming tasks logic.
"""

Reply:
Did add a comment. Did not include anything about locking though as it's not relevant for the consumer of this API.